### PR TITLE
Add vjp

### PR DESCRIPTION
**Pull request by sangyx**
Hi, I have added some files to complete `vjp`. These files come from [autodidact](https://github.com/mattjj/autodidact) and [autograd](https://github.com/HIPS/autograd). I realized that we may have to modify some existing code to implement it. So I make this pr first. These codes have not changed because I want to discuss with you how to modify.
There are several important differences between existing code and added files：

1. Basic differential function. This is the biggest difference. The existing diff functions take `x`, `x[0]` is the value, and `x[1]` is the diff. While the `vjp` diff functions take an additional parameter, the result of the function.
2.  We calculate all derivatives at once. Autograd only calculates one specified gradient at a time.
3. The derivation process is different. This is mainly due to the additional parameters that need to be passed on the first point.

I think we can solve these problems by rewriting or adding some code. For example, we can first check the difference between `compute_diff_tree` and `trace` and then decide how to modify it.

### CODE REVIEW

**Comment by hameerabbasi** [comment]
Please mark this PR as ready to review once it is done, I feel it is very much in flux and any review I give on it will soon be out of date. 😄

**Comment by sangyx** [comment]
OK. `VIP` is the most important part. I expect to make the code run through this week, then we still need to spend some time to test, review, and solve bugs.

**Comment by hameerabbasi** [comment]
That sounds like a decent timeline to me. 😄

**Comment by sangyx** [comment]
@hameerabbasi 
Can you help me see the latest pr code? Every time I run the last line(`print(y.diff(x1))`) of the [test.py](https://github.com/Quansight-Labs/udiff/pull/19/commits/97576a9d55070223865c4b7e7726b44ffb702c5b#diff-5a0c21c58f9ff9393621d3f108eb1b95), I get the error: `__ua_function__() missing 1 required positional argument:'kw'`

File: src/udiff/_diff_array.py [hunk 1]
```diff
@@ -1,139 +1,117 @@
...
+                diffs = self._vjp(grad_variables)
+                for pn, p in self._parents:
+                    p.backward(diffs[pn], self._visit)
+
+    def backward_jacobian(self):
+        """Backpropagation.
+        Traverse computation graph backwards in topological order from the end node.
+        For each node, compute local Jacobian contribution and accumulate.
+        """
+        pass
```
Comment (hameerabbasi) [inline src/udiff/_diff_array.py]: Maybe raise `NotImplementedError` here.
```suggestion
        raise NotImplementedError
```

File: src/udiff/_uarray_plug.py [hunk 2]
```diff
@@ -109,33 +84,20 @@ def __ua_convert__(value, dispatch_type, coerce):
             with udiff.SKIP_SELF:
                 return DiffArray(np.asarray(value))
 
-        return NotImplemented
```
Comment (hameerabbasi) [inline src/udiff/_uarray_plug.py]: Why this change?

File: src/udiff/_vjp_core.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            except KeyError:
+                raise NotImplementedError(
+                    "VJP of {} wrt argnums 0, 1 not defined".format(fun.__name__)
+                )
+            vjp_0 = vjp_0_fun(ans, *args, **kwargs)
+            vjp_1 = vjp_1_fun(ans, *args, **kwargs)
+            return lambda g: (vjp_0(g), vjp_1(g))
+        else:
+            vjps = [vjps_dict[argnum](ans, *args, **kwargs) for argnum in argnums]
+            return lambda g: (vjp(g) for vjp in vjps)
```
Comment (hameerabbasi) [inline src/udiff/_vjp_core.py]: I believe all these cases can be merged.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
All of this looks pretty awesome, @sangyx! I'd suggest you remove the commented code, and instead save it in a new branch or somewhere similar. Only a couple of minor comments.

**Comment by hameerabbasi** [comment]
Also, can you test something where the same input appears twice, e.g. `x ** x` or `x * x`.

File: src/udiff/_uarray_plug.py (hunk 2 above)
Comment (sangyx) [reply src/udiff/_uarray_plug.py]: Because in the NoGradBackend, we need to return the raw value of `np.array`.

Comment (hameerabbasi) [reply src/udiff/_uarray_plug.py]: What problems do you run into without removing this line? If not removed, it would return `None` which isn't an array anyway. In which case we should raise an error.

Comment (sangyx) [reply src/udiff/_uarray_plug.py]: When we set NoGradBackend, the `coerce` is `False` and the `value` is a `np.array`. Then we need to return the raw value, not a `DiffArray`. 
The code in the `backward` will go wrong without removing this line.

Comment (hameerabbasi) [reply src/udiff/_uarray_plug.py]: If we return `NotImplemented`, it will go to the next backend anyway, which will be a `np.ndarray`.

Comment (sangyx) [reply src/udiff/_uarray_plug.py]: @hameerabbasi You can test it by the code:
```python
import uarray as ua
import unumpy as np
import unumpy.numpy_backend as NumpyBackend
import udiff

if __name__ == "__main__":
    with ua.set_backend(NumpyBackend), ua.set_backend(udiff, coerce=True):
        x1 = np.array([[2, 2], [2, 2]])
        x2 = np.array([[5, 5], [5, 5]])
        y1 = np.log(x1) + x1 * x2 - np.sin(x2)
        print(y1)
        y1.backward()
        print(x1.diff)
        print(x2.diff)
```
It will go wring without removing this line.

**Comment by hameerabbasi** [comment]
This is in, thanks @sangyx!
